rviz_default_plugins/src/rviz_default_plugins/displays/map/palette_builder.cpp

```diff
@@ -0,0 +1,128 @@
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "rviz_default_plugins/displays/map/palette_builder.hpp"
+
+#include <cstring>
+#include <memory>
+#include <vector>
+
+namespace rviz_default_plugins
+{
+namespace displays
+{
+
+PaletteBuilder::PaletteBuilder()
+{
+  palette_ = std::vector<unsigned char>(256 * 4, 0);
+}
+
+std::shared_ptr<PaletteBuilder> PaletteBuilder::setColorForIllegalPositiveValues(
+  unsigned char r, unsigned char g, unsigned char b)
+{
+  for (unsigned char i = 101; i <= 127; i++) {
+    setColorForValue(i, r, g, b, 255);
+  }
+  return shared_from_this();
+}
+
+std::shared_ptr<PaletteBuilder> PaletteBuilder::setRedYellowColorsForIllegalNegativeValues()
+{
+  for (unsigned char i = 128; i <= 254; i++) {
+    // set shades from red to yellow
+    setColorForValue(i, 255, (255 * (i - 128)) / (254 - 128), 0, 255);
+  }
+  return shared_from_this();
+}
+
+std::shared_ptr<PaletteBuilder> PaletteBuilder::setColorForLegalNegativeValueMinusOne(
+  unsigned char r, unsigned char g, unsigned char b)
+{
+  setColorForValue(255, r, g, b, 255);
+  return shared_from_this();
+}
+
+std::shared_ptr<PaletteBuilder> PaletteBuilder::setColorForValue(
+  unsigned char palette_position,
+  unsigned char r, unsigned char g, unsigned char b, unsigned char alpha)
+{
+  palette_[4 * palette_position] = r;
+  palette_[4 * palette_position + 1] = g;
+  palette_[4 * palette_position + 2] = b;
+  palette_[4 * palette_position + 3] = alpha;
+  return shared_from_this();
+}
+
+std::vector<unsigned char> PaletteBuilder::buildPalette()
+{
+  return palette_;
+}
+
+std::vector<unsigned char> makeMapPalette()
+{
+  auto palette_builder = std::make_shared<PaletteBuilder>();
+  for (unsigned char i = 0; i <= 100; i++) {
+    // Standard gray map palette values
+    unsigned char v = 255 - (255 * i) / 100;
+    palette_builder->setColorForValue(i, v, v, v, 255);
+  }
+  return palette_builder->setColorForIllegalPositiveValues(0, 255, 0)
+         ->setRedYellowColorsForIllegalNegativeValues()
+         ->setColorForLegalNegativeValueMinusOne(0x70, 0x89, 0x86)
+         ->buildPalette();
+}
+
+std::vector<unsigned char> makeCostmapPalette()
+{
+  auto palette_builder = std::make_shared<PaletteBuilder>();
+  palette_builder->setColorForValue(0, 0, 0, 0, 0);
+  for (unsigned char i = 1; i <= 98; i++) {
+    unsigned char v = (255 * i) / 100;
+    palette_builder->setColorForValue(i, v, 0, 255 - v, 255);
+  }
+  palette_builder->setColorForValue(99, 0, 255, 255, 255);  // obstacle values in cyan
+  palette_builder->setColorForValue(100, 255, 0, 255, 255);  // lethal obstacle values in purple
+
+  return palette_builder->setColorForIllegalPositiveValues(0, 255, 0)
+         ->setRedYellowColorsForIllegalNegativeValues()
+         ->setColorForLegalNegativeValueMinusOne(0x70, 0x89, 0x86)
+         ->buildPalette();
```

Comment:
This indentation is strange to me. I would have expected it to either be aligned with the `->` on the first line or only have one extra indention (two spaces) from the `return`.

I guess this is what uncrustify did?

Reply:
Yes, I would have preferred the alignment with `->`, but it wouldn't let me do it.